ad5592r: Initial driver for ad5592r/ad5593r

mirceacaprioru opened the pull request:
The ad5592r/ad5593r has 8 I/Ox pins that can be independently configured as
digital-to-analog converter outputs (DAC), analog-to-digital converter
inputs (ADC) or digital inputs.

Signed-off-by: Mircea Caprioru <mircea.caprioru@analog.com>

stefpopa commented on drivers/ad5592r/ad5592r-base.c:
```diff
@@ -0,0 +1,291 @@
+/***************************************************************************//**
```
It would be nice to have a link to the datasheet in the commit message.

stefpopa commented on drivers/ad5592r/ad5592r.c:
```diff
@@ -0,0 +1,227 @@
+	int32_t ret;
+	uint16_t spi_msg_nop = 0; /* NOP */
+
+	ret = spi_write_and_read(dev->spi, (uint8_t *)&spi_msg_nop, sizeof(spi_msg_nop));
```
this line is clearly over 80 columns; please check this everywhere in your code

stefpopa commented on drivers/ad5592r/ad5592r-base.c:
```diff
@@ -0,0 +1,291 @@
+int32_t ad5592r_set_channel_modes(struct ad5592r_dev *dev)
+{
+	const struct ad5592r_rw_ops *ops = dev->ops;
+	uint32_t ret;
```
int32_t maybe?

stefpopa commented on drivers/ad5592r/ad5592r.c:
```diff
@@ -0,0 +1,227 @@
+	if (ret < 0)
+		return ret;
+
+	ret = ad5592r_set_channel_modes(dev);
```
you can probably return directly

stefpopa commented on drivers/ad5592r/ad5592r-base.c:
```diff
@@ -0,0 +1,291 @@
+	if (ret < 0)
+		return ret;
+
+	ret = dev->ops->reg_write(dev, AD5592R_REG_GPIO_IN_EN, dev->gpio_in);
```
return directly

stefpopa commented on drivers/ad5592r/ad5592r-base.c:
```diff
@@ -0,0 +1,291 @@
+		return FAILURE;
+
+	for (i = 0; i < dev->num_channels; i++) {
+		switch (dev->channel_modes[i]) {
```
maybe a lookup table would be a more elegant solution here

stefpopa commented on drivers/ad5592r/ad5593r.c:
```diff
@@ -0,0 +1,226 @@
+	if (ret < 0)
+		return ret;
+
+	ret = ad5592r_set_channel_modes(dev);
```
return directly

commodo commented on drivers/ad5592r/ad5592r-base.c:
```diff
@@ -0,0 +1,291 @@
+	else
+		dev->gpio_val &= ~BIT(offset);
+
+	dev->ops->reg_write(dev, AD5592R_REG_GPIO_SET, dev->gpio_val);
```
why not return int ?
`return dev->ops->reg_write(dev, AD5592R_REG_GPIO_SET, dev->gpio_val);`
?

commodo commented on drivers/ad5592r/ad5592r.h:
```diff
@@ -0,0 +1,53 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#define AD5592R_GPIO_READBACK_EN	BIT(10)
```
maybe add include guards here ?

commodo commented on drivers/ad5592r/ad5592r-base.h:
```diff
@@ -0,0 +1,111 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include "stdint.h"
```
add include guards ?

commodo commented on drivers/ad5592r/ad5593r.h:
```diff
@@ -0,0 +1,45 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+int32_t ad5593r_write_dac(struct ad5592r_dev *dev, uint8_t chan, uint16_t value);
```
include guards ?

adrimbarean commented on drivers/ad5592r/ad5592r.h:
```diff
@@ -0,0 +1,53 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
```
The header depends on 'stdint.h' and 'ad5592r-base.h' and might not build in certain situations if these files are not included.

dbogdan replied:
That's right. Make the header self-sufficient - it should be possible to standalone compile each header.

mirceacaprioru commented:
changelog V1 -> V2

- added include guards
- made headers self sufficient
- returned directly where possible
- verified column width to be maximum 80
- added wrapper for reg_write and read
- added links to datasheets to commit message

mirceacaprioru commented:
Changelod V2 -> V3 
- changed so that all returns are verified with if (ret < 0)

mirceacaprioru commented:
if there are no objections will merge later today.
